File: src/pneuma_core/models/memory.py
"""Memory data models: EpisodicMemory, SemanticMemory."""

from dataclasses import dataclass, field
from datetime import datetime


def _validate_range(value: float, name: str, min_val: float, max_val: float) -> float:
    if not (min_val <= value <= max_val):
        raise ValueError(f"{name} must be between {min_val} and {max_val}, got {value}")
    return value


@dataclass(frozen=True)
class EpisodicMemory:
    """エピソード記憶: 具体的な出来事の記録.

    emotional_valence: -1.0〜1.0（不快〜快）
    importance: 0.0〜1.0
    embedding: 1536次元ベクトル（None = 未生成）
    """

    id: str
    character_id: str
    content: str
    timestamp: datetime
    emotional_valence: float
    importance: float
    conversation_id: str | None = None
    embedding: list[float] | None = None

    def __post_init__(self) -> None:
        _validate_range(self.emotional_valence, "emotional_valence", -1.0, 1.0)
        _validate_range(self.importance, "importance", 0.0, 1.0)


@dataclass(frozen=True)
class SemanticMemory:
    """意味記憶: 汎化された知識.

    confidence: 0.0〜1.0（裏付けエピソード数で増加）
    """

    id: str
    character_id: str
    content: str
    confidence: float
    source_episode_ids: list[str] = field(default_factory=list)
    embedding: list[float] | None = None

    def __post_init__(self) -> None:
        _validate_range(self.confidence, "confidence", 0.0, 1.0)

File: akasha/contracts/Memory.ts
// Akasha Contract — Memory models
//
// Python 一次ソース: src/pneuma_core/models/memory.py
//
// 二重記憶システム (Episodic + Semantic) のデータ表現。

export interface EpisodicMemory {
  id: string;
  character_id: string;
  /** 出来事の自然言語表現。 */
  content: string;
  /** ISO 8601 形式の datetime 文字列 (Python 側は datetime)。 */
  timestamp: string;
  /** [-1, 1]。不快 → 快。 */
  emotional_valence: number;
  /** [0, 1]。重要度。 */
  importance: number;
  conversation_id?: string | null;
  /** 1536 次元程度の埋め込みベクトル (未生成時は null)。 */
  embedding?: number[] | null;
}

export interface SemanticMemory {
  id: string;
  character_id: string;
  /** 汎化された知識・パターンの自然言語表現。 */
  content: string;
  /** [0, 1]。裏付けエピソード数で増加する信頼度。 */
  confidence: number;
  /** 統合元のエピソード ID 集合。 */
  source_episode_ids: string[];
  embedding?: number[] | null;
}
